ports/models.py

```diff
@@ -62,13 +297,113 @@ class Builder(models.Model):
+            models.Index(fields=['port_name', 'builder_name']),
+            models.Index(fields=['-time_start']),
+            models.Index(fields=['port_name']),
+            models.Index(fields=['status'])
+        ]
+
+    @classmethod
+    def populate(cls):
+        builders = Builder.objects.values_list('name', flat=True)
+        url_prefix = 'https://build.macports.org'
+
+        def get_url_json(builder_name, build_number):
+            return '{}/json/builders/ports-{}-builder/builds/{}'.format(url_prefix, builder_name, build_number)
+
+        def get_url_build(builder_name, build_number):
+            return '{}/builders/ports-{}-builder/builds/{}'.format(url_prefix, builder_name, build_number)
+
+        def get_data_from_url(url):
+            gcontext = ssl.SSLContext()
+            with urllib.request.urlopen(url, context=gcontext) as u:
+                data = json.loads(u.read().decode())
+            return data
+
+        def get_build_properties(array):
+            properties = {}
+            for prop in array['properties']:
+                properties[prop[0]] = prop[1]
+            return properties
+
+        def get_build_steps(array):
+            steps = {}
+            for step in array['steps']:
+                steps[step['name']] = step
+            return steps
+
+        def return_summary(builder_name, build_number, build_data):
+            data = {}
+
+            properties = get_build_properties(build_data)
+            port_name = properties['portname']
+            status = ' '.join(build_data['text'])
+            time_start = build_data['times'][0]
+            time_build = float(build_data['times'][1]) - float(build_data['times'][0])
+
+            data['name'] = port_name
+            data['url'] = get_url_build(builder_name, build_number)
+            data['watcher_id'] = properties['triggered_by'].split('/')[6]
+            data['watcher_url'] = properties['triggered_by']
+            data['status'] = status
+            data['builder'] = builder_name
+            data['buildnr'] = build_number
+            data['time_start'] = str(datetime.datetime.fromtimestamp(int(float(time_start)), tz=datetime.timezone.utc))
+            data['buildtime'] = str(
+                datetime.timedelta(seconds=int(float(time_build)))) if time_build is not -1 else None
+
+            return data
+
+        def load_database(data):
+            builder = Builder.objects.get(name=data['builder'])
+            build = BuildHistory()
+            build.port_name = data['name']
+            build.status = data['status']
+            build.build_id = data['buildnr']
+            build.time_start = data['time_start']
+            build.time_elapsed = data['buildtime']
+            build.builder_name = builder
+            build.build_url = data['url']
+            build.watcher_url = data['watcher_url']
+            build.watcher_id = data['watcher_id']
+            build.save()
+
+        fulldata = {}
```

Comment:
Maybe move it to just before the loop `for buildername in builders:`.

Reply:
This is needed by the function `add_to_summary`.

Reply:
Then I think it would be better to return it from `add_to_summary`.

Reply:
Thank you, probably `full_data` and `add_to_summary` were both redundant.